OSM DATA 3D - Visualisation des données avec Giro3D, avec suivi des perf… (5/5) (#1284)
geotribu/website: content/articles/2025/2025-03-10_osm-data-3D-05-visulisation-3d.md

---
title: "OSM Data : visualisation des données 3D"
subtitle: OSM Data 5/5
authors:
    - Karl TAYOU
    - Romain LATAPIE
categories:
    - article
comments: true
date: 2025-03-31
description: "OSM DATA : visualisation des données avec Giro3D, avec suivi des performances"
icon: material/video-3d

> image: https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2025/osm_data/article_5/miniature.png

Reviewer: Changer pour une image qui corresponde aux dimensions attendues sinon c'est moche :

![image](https://github.com/user-attachments/assets/cb6d06ca-131b-4f73-a8ec-98be65aa4501)

license: default
robots: index, follow
tags:
    - WebGL
    - Three.js
    - Giro3D
    - OpenLayers

>     - Modélisation 3D

Reviewer: ordre A-z stp

---

# OSM DATA V2 : La visualisation en 3D d'OpenStreetMap

:calendar: Date de publication initiale : {{ page.meta.date | date_localized }}

## Introduction

> Dans les articles précédents, nous avons présenté les étapes de modélisation architecturale 3D à partir des données d'OpenStreetMap. Ce dernier article présente la méthode de visualisation des données 3D sur le web à l'aide de [Giro3D](https://giro3d.org/giro3d.html).
> 
> Pourquoi Giro3D ? Giro3D est un *framework* permettant de visualiser tout type de données géospatiales, en 2D, 2.5D ou 3D. Le moteur cartographique 2D est basé sur [OpenLayers](https://openlayers.org/), le moteur 3D repose sur [Three.js](https://threejs.org/), le projet est *opensource* et se veut *communautaire*, même s'il est actuellement principalement maintenu par [Oslandia](https://oslandia.com/). C'est pour le caractère *opensource*, la complémentarité avec OpenLayers, le développement régulier du produit et la réactivité du support qu'on a choisi cette technologie pour OSM DATA.

Reviewer: J'aime les images.

```suggestion
Dans les articles précédents, nous avons présenté les étapes de modélisation architecturale 3D à partir des données d'OpenStreetMap. Ce dernier article présente la méthode de visualisation des données 3D sur le web à l'aide de [Giro3D](https://giro3d.org/giro3d.html).

![logo Giro3D](https://cdn.geotribu.fr/img/logos-icones/logiciels_librairies/giro3d_rectangle.svg){: .img-center loading=lazy }

Pourquoi Giro3D ? Giro3D est un *framework* permettant de visualiser tout type de données géospatiales, en 2D, 2.5D ou 3D. Le moteur cartographique 2D est basé sur [OpenLayers](https://openlayers.org/), le moteur 3D repose sur [Three.js](https://threejs.org/), le projet est *opensource* et se veut *communautaire*, même s'il est actuellement principalement maintenu par [Oslandia](https://oslandia.com/). C'est pour le caractère *opensource*, la complémentarité avec OpenLayers, le développement régulier du produit et la réactivité du support qu'on a choisi cette technologie pour OSM DATA.
```

Comment représenter des milliers de données 3D avec des textures sur un navigateur web avec Giro3D afin que cela soit le plus fluide possible ? Comment mesurer la performance d'une telle application et identifier les différentes étapes consommatrices de ressources ? Quelle(s) méthode(s) utiliser pour permettre une navigation fluide ?

> ## Le suivi et l'optimisation de la performance de visualisation

Reviewer: ```suggestion
## Suivi et optimisation des performances de visualisation
```

Pour évoquer les performances de la solution, la notion de *frame* est importante. Dans le secteur multimédia, le terme *frame* représente une image, l’enchaînement d'images permet de créer une vidéo et le nombre d'images présentées successivement en une seconde représente le taux de rafraîchissement exprimé soit en Hertz, soit en nombre d'images par secondes (*frames per second*). Cette notion de *frame* est importante pour la suite car elle permet d'estimer les performances de la solution mise en en place.

> En visualisation 3D, on estime qu'un taux de rafraîchissement performant doit être de 60 *frames per second (fps)*. Par exemple, la Playstation 4 *Slim* permet un affichage à 30 fps alors que la Playstation 4 *pro* permet un affichage à 60 fps : expérience vécue dans FIFA, une frappe reçue de [Tammy Abraham](https://cdn.futwiz.com/assets/img/fc24/social/cards/604_pos.png?v=225516.png) dans la lucarne de Steve Mandanda est deux fois plus fluide sur la Playstation 4 *Pro* que sur la Playstation 4 *Slim*.

Reviewer: J'aime les images et je trouve que cette partie mériterait une illustration explicative (schéma ou autre à trouver) ou simplement illustrative 

![](https://images.huehd.com/wp-content/uploads/2021/12/The_Horse_in_Motion_high_res-1024x631.jpg) par exemple ?

Author: je peine à trouver une illustration qui représenterai ce que c'est le FPS. L'image des chevaux, représente 12 frames permettant de voir un galop de cheval c'est bien ca ? J'ai l’impression que l'image illustre plus le concept de "frame" que celui du FPS, il y'a pas une notion de temps sur l'image.

Reviewer: Il y a forcément une notion de temps mais je comprends ce que tu dis. Pas grave :)

Plus sérieusement et dans le cas d'OSM DATA, si on considère une scène avec 1000 objets 3D, le moteur de rendu doit être capable d'interpréter et d'afficher les 1000 objets 60 fois par seconde.

Le défi de notre application est donc d'être capable de générer des scènes 3D avec un taux de rafraîchissement suffisant pour proposer une expérience utilisateur agréable, c'est-à-dire sans *bugs* (défaut de fonctionnement) et sans *lags* (retard/décalage d'images). Pour cela, étant donné que Giro3D est basé sur Three.js, lui-même basé sur [WebGL](https://get.webgl.org/), il est recommandé une fréquence d'affichage de [60 fps](http://www.opengl-tutorial.org/fr/miscellaneous/an-fps-counter/), soit 1 image toutes les 17 millisecondes.

> Ce suivi est rendu possible dans Giro3D à l'aide d'un utilitaire directement intégré (`import Inspector from '@giro3d/giro3d/gui/Inspector.js';`) et qui n'est pas encore inclus dans la documentation. :trollface:

Reviewer: @sguimmara @autra vous perdez un point de karma documentariste et la portée de votre pouvoir RTFM est réduite de 20%. :game_die:

Reviewer: Tiens on peut ajouter de la coloration syntaxique inline https://squidfunk.github.io/mkdocs-material/reference/code-blocks/?h=highligh#highlighting-inline-code-blocks

```suggestion
Ce suivi est rendu possible dans Giro3D à l'aide d'un utilitaire directement intégré (`#!typescript import Inspector from '@giro3d/giro3d/gui/Inspector.js';`) et qui n'est pas encore inclus dans la documentation. :trollface:
```

Reviewer 2: Par contre, notre pouvoir de lancer le passif-agressif "contribution welcome" est augmenté de 50%! Et je gagne en combo 1 pts mauvaise foi. Échange équitable ;-)

En exécutant la commande JavaScript, on obtient un graphique représentant la durée de visualisation (en ordonnée) pour chaque image définie au cours du temps (en abscisse).

![Inspecteur de Giro3D afin de suivre le taux de rafraichissement](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2025/osm_data/article_5/inspecteur_giro3d.png){: .img-center loading=lazy }

> Les principales étapes de calcul d'une image dans notre scène sont les suivantes, elles sont exécutées au niveau du *Central Processing Unit (CPU)* :

Author: Dans OSM DATA, il y'a pas (encore) d'ombrages; La lumière , elle est très peu utilisée et que dans les shader (GPU).

1. Application des transformations géométriques des objets (position, rotation, échelle) engendré par le déplacement dans la scène 3D
2. Application des lumières et des ombres en fonction des nouvelles conditions de position et d'orientation de la *pose* considérée
3. Génération à la volée des objets 3D présents dans le cône de vue défini.

> Une fois les calculs effectués, le rendu 3D de la scène est généré par la réalisation d'appels au *Global Processing Unit (GPU)*.

Reviewer 3: ```suggestion
Une fois les calculs effectués, le rendu 3D de la scène est généré par la réalisation d'appels au *Graphics Processing Unit (GPU)*.
```

> L'ensemble de ces étapes peuvent être chronophages et excrément énergivores si on considère une réalisation du processus pour chaque objet affiché, voici un résultat obtenu avec mon ordinateur qui a perdu quelques années de vie juste pour vous lecteurs de Geotribu :

Reviewer 3: ```suggestion
L'ensemble de ces étapes peuvent être chronophages et extrêmement énergivores si on considère une réalisation du processus pour chaque objet affiché. Voici un résultat obtenu avec mon ordinateur qui a perdu quelques années de vie juste pour vous lecteurs de Geotribu :
```

Reviewer: 💩

Author: décidément, on dirait qu'il y'avait un truc correcteur bizarre sur mon ordi. Merci 🙏

> L'ensemble de ces étapes peuvent être chronophages et extrêmement énergivores si on considère une réalisation du processus pour chaque objet affiché. Voici un résultat obtenu avec mon ordinateur qui a perdu quelques années de vie juste pour vous lecteurs de Geotribu :

Reviewer: ```suggestion
L'ensemble de ces étapes peut être chronophage et extrêmement énergivore si on considère une réalisation du processus pour chaque objet affiché. Voici un résultat obtenu avec mon ordinateur qui a perdu quelques années de vie juste pour vous lecteurs de Geotribu :
```

![Utilitaire de *monitoring* pour une génération séquentielle d'images](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2025/osm_data/article_5/inspecteur_giro3d_non_optimise.png){: .img-center loading=lazy }

> Les temps de génération sont excrément longs, variant de 1.5 secondes à ... 8 secondes ! :dizzy_face:

Reviewer 3: ```suggestion
Les temps de génération sont extrêmement longs, variant de 1,5 secondes à ... 8 secondes ! :dizzy_face:
```

> ![PC Killer](https://media.tenor.com/hEwfEcj2R60AAAAM/laptop-smoking.gif){: .img-center loading=lazy }

Reviewer: ```suggestion
![PC Killer](https://cdn.geotribu.fr/img/articles-blog-rdp/gifs/laptop-smoking.gif){: .img-center loading=lazy }
```

Afin de réduire ce temps de génération, plusieurs actions peuvent être mises en place :

- Paramétrer Giro3D afin de ne pas afficher des objets qui ne sont pas visibles sur l'écran de l'utilisateur. On utilise pour cela une fonction [Frustum](https://threejs.org/docs/index.html?q=frus#api/en/math/Frustum) permettant de sélectionner un morceau de la scène à afficher. Cette action permet de réduire significativement le temps de traitement de toutes les étapes, elle est mise en place dans OSM DATA.

> - Diminuer au maximum le volume des données envoyées au GPU. Cette étape peut être mise en place en limitant la répétition de coordonnées, de textures, ... mais elle limite les potentielles évolutions de la plateforme, cette action n'est pas mise en place dans OSM DATA.

Reviewer 4: ```suggestion
- Diminuer au minimum le volume des données envoyées au GPU. Cette étape peut être mise en place en limitant la répétition de coordonnées, de textures... mais elle limite les potentielles évolutions de la plateforme, cette action n'est pas mise en place dans OSM DATA.
```

Reviewer 2: Faire quelque chose au maximum, ça se dit non ?

Reviewer 4: Des éléments de contexte ici https://www.antidote.info/fr/blogue/enquetes/reduire-au-maximum
--> possible de changer la tournure de la phrase pour éviter le minimum/maximum : minimiser le volume de données. Ou alors optimiser qui irait aussi ici comme on cherche la performance.

Author: make sens

> - Générer une nouvelle scène si et seulement aucune modification de *pose* n'est intervenue depuis 200 millisecondes, cette action est mise en place dans OSM DATA.

Reviewer 4: ```suggestion
- Générer une nouvelle scène si et seulement si aucune modification de *pose* n'est intervenue depuis 200 millisecondes, cette action est mise en place dans OSM DATA.
```

> - Tuiler la scène sous la forme d'une grille de tuiles carrées contigues de 512 pixels * 512 pixels. Cette action permet d'accélérer considérablement l'affichage, les objets présents dans une tuile ("enfants") peuvent être fusionnés en seul objet rattaché à la tuile "parent". Les changements de points de vue sont affectés aux entités de type "parent", moins nombreuses que les entités de type "enfant", assurant un affichage optimisé. Cette action est mise en place dans OSM DATA.

Reviewer 4: ```suggestion
- Tuiler la scène sous la forme d'une grille de tuiles carrées contigues de 512 pixels * 512 pixels. Cette action permet d'accélérer considérablement l'affichage, les objets présents dans une tuile ("enfants") peuvent être fusionnés en un seul objet rattaché à la tuile "parent". Les changements de points de vue sont affectés aux entités de type "parent", moins nombreuses que les entités de type "enfant", assurant un affichage optimisé. Cette action est mise en place dans OSM DATA.
```

- Extruder directement les objets en 3D côté serveur (avec un cache). Cette action n'est pas mise en place dans OSM DATA (pour l'instant :trollface:)

Cette liste d'optimisations n'est pas exhaustive. En moyenne, avec un ordinateur "standard" (8Go RAM, processeur Intel core-i7), l'application dispose d'un taux de rafraîchissement supérieur à 50 fps.

Dans la prochaine partie, nous approfondissons la méthode de tuilage sous forme de grille.

## Tuilage de la scène

> Au fil des interactions avec la scène, les données sont récupérées depuis le serveur cartographique en fonction de la cône visible et uniquement au niveau de *zoom* 16 (zoom à partir duquel les bâtiments extrudés s'affichent - choix arbitraire). La grille est générée à ce niveau de *zoom* avec une taille de 512 pixels * 512 pixels. Afin de limiter les ressources en CPU et GPU, chaque tuile doit pouvoir :

Reviewer 3: ```suggestion
Au fil des interactions avec la scène, les données sont récupérées depuis le serveur cartographique en fonction du cône visible et uniquement au niveau de *zoom* 16 (zoom à partir duquel les bâtiments extrudés s'affichent - choix arbitraire). La grille est générée à ce niveau de *zoom* avec une taille de 512 pixels * 512 pixels. Afin de limiter les ressources en CPU et GPU, chaque tuile doit pouvoir :
```

- Être chargée une seule fois avec un seul objet fusionné (ensemble des objets présents au sein de l'emprise de la tuile) : cette contrainte permet d’éviter la répétition d'opérations d'extrusion et de fusion géométrique.

> - Être affichée dès que tous les objets présents au sein de la tuile sont générés

Reviewer 4: ```suggestion
- Être affichée dès que tous les objets présents au sein de la tuile sont générés.
```

Pour réussir ce double objectif, la bibliothèque OpenLayers dispose d'une classe de type [VectorTile](https://openlayers.org/en/v6.15.1/apidoc/module-ol_source_VectorTile-VectorTile.html). Cette classe permet de récupérer directement les données depuis le serveur en les affectant à une grille dont la dimension est spécifiée.

```javascript title="Initialisation de notre source OpenLayers"
const buildingTileSource = new VectorTileSource({
        url: "URL_SERVEUR_CARTO_BATIMENT",
        tileGrid: createXYZ({ tileSize: 512 }) // Spécification de la dimension de notre tuile
});
```

A chaque mouvement de la carte, toutes les tuiles du zoom 16 situées dans la zone visible par l'utilisateur sont récupérées : certaines sont déjà affichées dans la scène (pas de chargement nécessaire), les autres sont chargées pour afficher une scène complète.

```javascript title="Demande de chargement des tuiles"
import VectorRenderTile from "ol/VectorRenderTile";
import { TileCoord } from "ol/tilecoord";
import TileState from "ol/TileState"

// Tableau qui stockera toutes les tuiles à charger
const loadingTiles: VectorRenderTile[] = []

> // On parcourt toutes les coordonées (x,y,zoom) des tuiles du niveau 16 dans la zone visible  par l'utilisateur

Reviewer 4: ```suggestion
// On parcourt toutes les coordonnées (x,y,zoom) des tuiles du niveau 16 dans la zone visible  par l'utilisateur
```

// mapExtent : zone visible par l'utilisateur
buildingTileSource.tileGrid.forEachTileCoord(mapExtent, 16, (tileCoord: TileCoord) => {
    const zoom = tileCoord[0]
    const x = tileCoord[1]
    const y = tileCoord[2]

>     // On récupère notre tuile à partir de ses coordonées

Reviewer 4: ```suggestion
    // On récupère notre tuile à partir de ses coordonnées
```

    const tile = this.vectorTileSource.getTile(zoom, x, y, "RESOLUTION DE NOTRE SCENE", "EPSG:3857")
    // Si notre tuile a un statut IDLE, ce qu'elle est n'est pas encore chargée
    if (tile.getState() == TileState.IDLE) {
        // On demande à notre source de la charger depuis le serveur cartographique
        tile.getSourceTiles()
        loadingTiles.push(tile)
    }
})
```

> Lorsqu'une tuile est entièrement chargée dans la scène, il convient de l'identifier. Pour cela, la bibliothèque [RxJs](https://rxjs.dev/) facilite la gestion des changements, nous pouvons suivre efficacement l’état de chargement des données des tuiles et déclencher les actions nécessaires.

Reviewer 3: ```suggestion
Lorsqu'une tuile est entièrement chargée dans la scène, il convient de l'identifier. Pour cela, la bibliothèque [RxJs](https://rxjs.dev/) facilite la gestion des changements: nous pouvons suivre efficacement l’état de chargement des données des tuiles et déclencher les actions nécessaires.
```

> Lorsqu'une tuile est entièrement chargée dans la scène, il convient de l'identifier. Pour cela, la bibliothèque [RxJs](https://rxjs.dev/) facilite la gestion des changements: nous pouvons suivre efficacement l’état de chargement des données des tuiles et déclencher les actions nécessaires.

Reviewer 4: ```suggestion
Lorsqu'une tuile est entièrement chargée dans la scène, il convient de l'identifier. Pour cela, la bibliothèque [RxJs](https://rxjs.dev/) facilite la gestion des changements : nous pouvons suivre efficacement l’état de chargement des données des tuiles et déclencher les actions nécessaires.
```

```javascript title="Surveillance du chargement des tuiles afin de les traiter"

import { BehaviorSubject, interval } from 'rxjs';
import { map } from 'rxjs/operators';

// Fonction qui récupère l'état actuel des tuiles à charger qui sont normalement tous en IDLE (0) à l'initialisation
const getTileState = (): { tile: VectorRenderTile, state: number }[] => {
    return tilesToLoad.map((tile) => ({ tile, state: tile.getState() }));
};

// Un observable pour suivre les états des tuiles
// NB : Il ne se met pas à jour tout seul
const getTileState$ = new BehaviorSubject(getTileState());

// Abonnement périodique (500 ms) pour vérifier les changements d'état des tuiles
const updateTileStateSubscription = interval(500).pipe(
    map(() => {
        const previousStates = getTileState$.value;
        const currentStates = getTileState();

        // On filtre les tuiles dont l'état a changé

>         // En 500 ms plusieurs tuiles peuvent avoir changées

Reviewer 4: ```suggestion
        // En 500 ms plusieurs tuiles peuvent avoir changé
```

        const changedTiles = currentStates.filter((currentTileState, index) => {
            const previousTileState = previousStates[index];
            return currentTileState.state !== previousTileState.state;
        });

>         // On oublie pas de mettre à jour l'observable avec les nouveaux états

Reviewer 4: ```suggestion
        // On n'oublie pas de mettre à jour l'observable avec les nouveaux états
```

        // afin que dans 500 ms, le filtre précédent puisse détecter s'il y a des changements ;)
        getTileState$.next(currentStates);

        // On continue uniquement avec les tuiles dont l'etat a changé
        return changedTiles;
    })
).subscribe((changedTiles:{ tile: VectorRenderTile, state: number }[]) => {
    // On a nos tuiles à traiter
    changedTiles.forEach(({ tile:VectorRenderTile, state }) => {

        // la tuile Tile, n'attend plus qu'à être traitée...
    });
});

```

Pour les tuiles chargées, nous devons pour chacune d'elles extruder les objets, les fusionner et les ajouter à la scène. Cependant :

- La grille générée par OpenLayers n'est pas directement compatible avec Giro3D. Nous devons créer et gérer une grille spécifique à Giro3D en conservant les paramètres de celle d’OpenLayers (à savoir les emprises des tuiles).

> - Les coordonnées des objets issus du serveur sont en projection planimétrique [WGS84 Pseudo-Mercator (EPSG:3857)](https://epsg.io/3857). Pour optimiser les performances, les coordonnées doivent être modifiées et reliées à un système de coordonnées local propre à chaque tuile : elles ne sont plus absolues mais relative à l'origine définie pour la tuile considérée.

Reviewer 3: ```suggestion
- Les coordonnées des objets issus du serveur sont en projection planimétrique [WGS84 Pseudo-Mercator (EPSG:3857)](https://epsg.io/3857). Pour optimiser les performances, les coordonnées doivent être modifiées et reliées à un système de coordonnées local propre à chaque tuile : elles ne sont plus absolues mais relatives à l'origine définie pour la tuile considérée.
```

```javascript title="Création de la tuile sous Three.JS"

import {Group} from "three";

// Notre tuile OpenLayers déja chargée
const LoadedTile:VectorRenderTile
// Tableau des features réprésentant les objets de notre tuile
const features:Feature[]

// Coordonnées inférieur gauche de notre tuile OpenLayers
const LoadedTileBottomLeft = getBottomLeft(buildingTileSource.tileGrid.getTileCoordExtent(LoadedTile.getTileCoord()))

const x = LoadedTileBottomLeft[0]
const y = LoadedTileBottomLeft[1]

// Création de notre tuile Three.JS
const treeJsTile = new Group()
// Définition des coordonnées de la tuile Three.JS
treeJsTile.position.set(x,y,0)

// Pour chaque feature (objet), on ajuste les coordonnées
features.map((feature)=>{
    const polygon = feature.getGeometry()
    const newCoordinates = polygon.getLinearRings().map((ring)=>{
        return ring.getCoordinates().map((coordinate:Coordinate)=>{
            // Ajustement des coordonnées par rapport aux coin inférieur gauche de sa tuile
            return [
                coordinate[0] - x,
                coordinate[1] - y
            ]
        })

    })
    polygon.setCoordinates(newCoordinates)
})

> // On construit la géométrie unique pour tous nos objets : ceci utilise toutes les méthodes détaillées dans les deux articles précédent de cette série

Reviewer 4: ```suggestion
// On construit la géométrie unique pour tous nos objets : ceci utilise toutes les méthodes détaillées dans les deux articles précédents de cette série
```

const buildingsGeometry = buildBuildingsFromFeatures(features)
// Ajout de la tuile directement à notre scène Giro3D

> instance.add(treeJsTile)

Reviewer 2: Je crois que je ne comprends pas pourquoi tout cela est nécessaire (à partir de la ligne 96 de cet article) : giro3d supporte les sources xyz en vector tile directement. Il ne charge que les données dans le frustrum. Il sait même extrudé les données à la volée selon des métadonnées des geoms.  Pourquoi OSM Data a besoin de gérer le chargement lui-même ?

Author: Oui giro3D supporte bien les sources XYZ, mais ne permets pas d'extruder les toitures tel que décrit dans l'article 3/5 en fonction des attributs d'OpenStreeMap (je ne pense pas d'ailleurs que ce soit sa vocation), de gérer l'affichage de milliers d'objets (bâtiments+points+lignes) tout en utilisant des textures + shader plus ou moins complexe. 

En gros le besoin d'OSM DATA était d'avoir un rendu poussé tout en affichant de milliers d'objets, qui en fait un besoin spécifique. 

Quoi que la gestion/affichage des points tel que fait sur OSM DATA pourrait constituer une véritable couche au sens `layer` sous Giro3D :  une `Mesh` par tile, avec pour géométrie un `InstancedBufferGeometry`.  A réfléchir dans le sens ou cela complexifie beaucoup le code notemment avec la gestion d'un shader.

Reviewer 2: Ok merci de la clarification, c'est intéressant !

Ceci étant dit, peut-être qu'une intégration plus poussée dans giro3d pourrait simplifier les choses. Par ex, toute instance de giro3d expose l'événement 'object-created' sur lequel on peut s'accrocher pour modifier l'objet qui vient d'arriver du serveur. Potentiellement ça peut simplifier le cas d'usage ici (mais je n'ai pas testé)

Enfin bon, ça dépasse le cadre de cet article en tout cas...

Reviewer 2: > Par ex, toute instance de giro3d expose l'événement 'object-created' sur lequel on peut s'accrocher pour modifier l'objet qui vient d'arriver du serveur. Potentiellement ça peut simplifier le cas d'usage ici (mais je n'ai pas testé)

En tout cas, ça te permettrait de ne pas abandonner la gestion des tuiles, des protocols, des requêtes réseaux de giro3d parce que tu as besoin de modifier l'objet reçu. 

J'imagine que le code dans cet event listener serait grosso modo le code dans ce forEach (avec peut-être des adaptations):

```js
  changedTiles.forEach(({ tile:VectorRenderTile, state }) => {

        // la tuile Tile, n'attend plus qu'à être traitée...
    });
```
Vu ton cas d'usage, peut-être qu'un `tile-created` event qui n'existe pas encore (avant la génération du mesh) te serait d'ailleurs plus utile, je ne sais pas.

En tout cas n'hésite pas à nous suggérer (côté dépôt giro3d) des nouvelles API pour couvrir ton besoin!

Reviewer 2: carrément, j'ai commencé le thread.

```

Nous avons présenté dans cet article les spécificités de visualisation avec Giro3D dans OSM DATA : les processus de gestion, de traitement et d'affichage des données 3D doivent être bien déterminés et optimisés en amont afin d'assurer une visualisation fluide des données.

> 
> On espère que cette série d'articles sur les nouveautés d'OSM DATA vous a permis d'en savoir un peu plus sur les possibilités d'exploitation de la base de données d'OpenStreetMap. On espère vous avoir fait partagé notre appétence pour l'*opensource*, l'interopérabilité et l'interconnexion SIG / 3D / jumeaux numériques. Nous reviendrons écrire ici mais nous devons d'abord repartir nous entrainer à FIFA ! :soccer:
> 

Reviewer: J'ajouterais bien un niveau de titre pour marquer la fin de la série :

```suggestion

## Fin de série

On espère que cette série d'articles sur les nouveautés d'OSM DATA vous a permis d'en savoir un peu plus sur les possibilités d'exploitation de la base de données d'OpenStreetMap. On espère vous avoir fait partagé notre appétence pour l'*opensource*, l'interopérabilité et l'interconnexion SIG / 3D / jumeaux numériques. Nous reviendrons écrire ici mais nous devons d'abord repartir nous entrainer à FIFA ! :soccer:

```
